### DISCUSSION TITLE
API for Addons

### DISCUSSION THREAD

**Discussion by Explosif3005** [Luke100000/minecraft-comes-alive#837] [Ideas]
Hello I was thinking that there where some part of MCA that could be given more customisation for pack and addons maker. For exemple wich block is used by the building scanner to create new special house (to have support for modded block) and have the opportunity to add new custom special build, with potentially some handler for the AI of villager, for exemple have the ability to say that say a lectern and chair (from a mod) form a forum, that become a spot were villager can go and discuss probably improving the mood. And such little thing. I think it should be relatively simple and give addons maker an more easier time doing thing.

**Comment by Luke100000** [comment]
> wich block is used by the building scanner to create new special house

Supported, but not documented: https://github.com/Luke100000/minecraft-comes-alive/tree/1.19.2/common/src/main/resources/data/mca/building_types

I want to expand on "day of the week" as well as the concept of special days. I will keep data driven approaches in mind to support modding it.

**Reply by Explosif3005** [reply]
Is it possible to use tags in those json? If no is it possible to add it?

And suggestion:
Certain block or tags could also be forbidden from being a certain house to have the jobs. For exemple having multiple tnt in a inn stop it from being one (As it could be too explosive) and this kind of stuff. This can allow to restric certain building using certains blocks when building

**Reply by Explosif3005** [reply]
But still thanks for the answer

**Reply by Luke100000** [reply]
Yes, prefix them with # as used here: https://github.com/Luke100000/minecraft-comes-alive/blob/1.19.2/common/src/main/resources/data/mca/building_types/infirmary.json

Blacklisting sounds interesting, but I don't see many use cases. Explosive inns appear very specific.

**Reply by Explosif3005** [reply]
well for exemple there is the use of block such a cauldron for decoration purposes as such we do not want it to be used to determine the type of the building, could help like that to avoid certain confusion for the scanner with custom building

**Reply by Explosif3005** [reply]
And how should i go to add block or tag to already existing building. For compatibiliy for mods such as Conquest Reforged that add new block.

**Reply by Luke100000** [reply]
How would you make an infirmary then?

Iirc, I made it that datapacks overwrite building types.

**Reply by Explosif3005** [reply]
Overvrite completely?

**Reply by Luke100000** [reply]
I assume so

**Reply by Explosif3005** [reply]
I added a proposition for data driven modification of AI of Villager lower in the conversation

**Comment by Explosif3005** [comment]
Also when I find the time to truly sit down and write about it I have a proposition to add data driven task for AI

**Comment by Explosif3005** [comment]
I was inspired by other piece of data in MCA for that

**Comment by Explosif3005** [comment]
The Task for AI data driven should be divided into three part:
Location rules:
{Physical location -> a type of MCA Reborn house or a type of Structure
Block location -> a name of block and optionnal tag of block that have to be matched
Block rule -> Additionnal block data have to follow the rule (here i speak more of NBT data)
}
Activation rules:
{Villager: The data of the Villager has to mach the rules (data will also include the last task it as done)
Other Date: Other World Data such as the presence of a item of block at a location, or the time of the day.
}
Actions:
[ A quantity of action wich can range from interacting with blocks, producing resources, changing mood, moving or staying put, killing and attaking specific entity, going to a specific location, or going to the location of the task, moving items.... There should also be the ability to use special fonction set in the code to extract data from an item and use it in a following action. (For that look at tetra there should be something similar with the ability of weapon and tool. Such as the abilities of the shield
]
